coco/worker.py

```diff
@@ -74,41 +71,74 @@ def main_loop(endpoints, forwarder, coco_port, metrics_port, log_level):
-            except KeyError:
-                msg = f"endpoint /{endpoint_name} not found."
-                logger.debug(f"coco.worker: Received request to /{endpoint_name}, but {msg}")
-                await conn.execute(
-                    "rpush",
-                    f"{name}:res",
-                    json.dumps(Result(endpoint_name, result=None, error=msg).report()),
-                )
-                continue
 
-            if method != endpoint.type:
-                msg = (
-                    f"endpoint /{endpoint_name} received {method} request (accepts "
-                    f"{endpoint.type} only)"
-                )
-                logger.debug(f"coco.worker: {msg}")
-                await conn.execute(
-                    "rpush",
-                    f"{name}:res",
-                    json.dumps(Result(endpoint_name, result=None, error=msg).report()),
-                )
-                continue
+                if not request:
+                    request = None
+                else:
+                    try:
+                        request = json.loads(request)
+                    except json.JSONDecodeError as e:
+                        raise InvalidUsage(f"Invalid JSON payload: {request}") from e
+                    # Check that the requested endpoint exists
+                    if endpoint_name not in endpoints:
+                        msg = f"endpoint /{endpoint_name} not found."
+                        logger.debug(
+                            f"coco.worker: Received request to /{endpoint_name}, but {msg}"
+                        )
+                        raise InvalidPath(msg)
 
-            logger.debug(f"coco.worker: Calling /{endpoint.name}: {request}")
-            result = await endpoint.call(request)
+                endpoint = endpoints[endpoint_name]
+
+                # Check that it is being requested with the correct method
+                if method != endpoint.type and method not in endpoint.type:
+                    msg = (
+                        f"endpoint /{endpoint_name} received {method} request (accepts "
+                        f"{endpoint.type} only)"
+                    )
+                    logger.debug(f"coco.worker: {msg}")
+                    raise InvalidMethod(msg)
+
+                logger.debug(f"coco.worker: Calling /{endpoint.name}: {request}")
+                result = await endpoint.call(request)
+
+                # Transform any Result into a report so it can be serialised
+                if isinstance(result, Result):
+                    result = result.report()
+
+                code = 200
+
+            # Process a known exception source into a response
+            except CocoException as e:
+                result = e.to_dict()
+                code = e.status_code
+
+            # Unexpected exceptions are returned as HTTP 500 errors, and dump a
+            # traceback
+            except Exception as e:
```

Comment:
```suggestion
            except BaseException as e:
```

to include SystemExit, KeyboardInterrupt, GeneratorExit (https://docs.python.org/3/library/exceptions.html#exception-hierarchy)?

Reply:
Don't we want those exceptions to cause a genuine exit from the mainloop?

Reply:
Ah, yes. I was just thinking the client should get a reply instead of timing out. We could re-raise in that case.

Reply:
Fair enough. I can add a final clause that catches, replies, and re-raises any base exception. I think I need to remove the reply into a finally block